PersistentStorage/Sources/Data/Repository/PersistentStorage/MainPersistentStorage.swift

```diff
@@ -14,154 +13,144 @@ public class MainPersistentStorage: PersistentStorage {
+        valueKey: String,
+        userDefaults type: UserDefaultsType
     ) -> AnyPublisher<Void, PersistentStorageError> {
-        userDefaults.removeObject(forKey: valueKey)
-        log(
-            message: "✅ Successfully removed value with key {\(valueKey)}",
-            for: .debug
-        )
-        // This operation is always successful
-        return Just(())
-            .setFailureType(to: PersistentStorageError.self)
+        userDefaults(for: type)
+            .map { [weak self] userDefaults in
+                userDefaults.removeObject(forKey: valueKey)
+                self?.log(
+                    message: "✅ Successfully removed value with key {\(valueKey)} in userDefaults {\(type)}",
+                    for: .debug
+                )
+                // This operation is always successful
+                return ()
+            }
             .eraseToAnyPublisher()
     }
 
-    /// Method for reading a value with given key and type. Result is returned as a publisher.
-    ///
-    /// - Possible failures:
-    ///     - If type resolution with given parameter `valueType` fails, method will return `noValueFoundWithGivenType` failure.
-    ///     - If no value is associated with key given in parameter `valueKey` or its value is nil, the method will return `noValueFound` failure.
-    ///
-    /// - Parameters:
-    ///     - valueType: The type of value that you wish to read.
-    ///     - valueKey: Key that is assigned to a value that you wish to read.
-    public func readWithPublisher<T>(
+    public func read<T>(
         valueType: T.Type,
-        valueKey: String
+        valueKey: String,
+        userDefaults type: UserDefaultsType
     ) -> AnyPublisher<T, PersistentStorageError> {
-        do {
-            let persistedValue = try getPersistedValue(
-                valueType: valueType,
-                valueKey: valueKey
-            )
-            return Just(
-                persistedValue
-            )
-            .setFailureType(to: PersistentStorageError.self)
-            .eraseToAnyPublisher()
-        } catch {
-            guard let error = error as? PersistentStorageError else {
-                return Fail(
-                    error: PersistentStorageError.undefined
-                ).eraseToAnyPublisher()
+        userDefaults(for: type)
+            .flatMap { [weak self] userDefaults -> AnyPublisher<T, PersistentStorageError> in
+                guard let persistedValue = userDefaults.value(forKey: valueKey) as? T else {
+                    if userDefaults.value(forKey: valueKey) != nil {
+                        self?.log(
+                            message: "❌ Value with given key exists in userDefaults {\(type)}, but method is unable to parse the value with given type.",
+                            for: .failure
+                        )
+                        return Fail(error: PersistentStorageError.noValueFoundWithGivenType).eraseToAnyPublisher()
```

Comment:
No need to write enum name explicitly. Only `.noValueFoundWithGivenType` should be enough.